proposals/csharp-10.0/lambda-improvements.md

```diff
@@ -0,0 +1,200 @@
+﻿# Lambda improvements
+
+## Summary
+Proposed changes:
+1. Allow lambdas with attributes
+2. Allow lambdas with explicit return type
+3. Infer a natural delegate type for lambdas and method groups
+
+## Motivation
+Support for attributes on lambdas would provide parity with methods and local functions.
+
+Support for explicit return types would provide symmetry with lambda parameters where explicit types can be specified.
+Allowing explicit return types would also provide control over compiler performance in nested lambdas where overload resolution must bind the lambda body currently to determine the signature.
+
+A natural type for lambda expressions and method groups will allow more scenarios where lambdas and method groups may be used without an explicit delegate type, including as initializers in `var` declarations.
+
+Requiring explicit delegate types for lambdas and method groups has been a friction point for customers, and has become an impediment to progress in ASP.NET with recent work on [MapAction](https://github.com/dotnet/aspnetcore/pull/29878).
+
+[ASP.NET MapAction](https://github.com/dotnet/aspnetcore/pull/29878) without proposed changes (`MapAction()` takes a `System.Delegate` argument):
+```csharp
+[HttpGet("/")] Todo GetTodo() => new(Id: 0, Name: "Name");
+app.MapAction((Func<Todo>)GetTodo);
+
+[HttpPost("/")] Todo PostTodo([FromBody] Todo todo) => todo);
+app.MapAction((Func<Todo, Todo>)PostTodo);
+```
+
+[ASP.NET MapAction](https://github.com/dotnet/aspnetcore/pull/29878) with natural types for method groups:
+```csharp
+[HttpGet("/")] Todo GetTodo() => new(Id: 0, Name: "Name");
+app.MapAction(GetTodo);
+
+[HttpPost("/")] Todo PostTodo([FromBody] Todo todo) => todo);
+app.MapAction(PostTodo);
+```
+
+[ASP.NET MapAction](https://github.com/dotnet/aspnetcore/pull/29878) with attributes and natural types for lambda expressions:
+```csharp
+app.MapAction([HttpGet("/")] () => new Todo(Id: 0, Name: "Name"));
+app.MapAction([HttpPost("/")] ([FromBody] Todo todo) => todo);
+```
+
+## Attributes
+Attributes may be added to lambda expressions.
+```csharp
+f = [MyAttribute] x => x;          // [MyAttribute]lambda
```

Comment:
This is the form that I think should not be allowed because it could plausibly mean either `[MyAttribute] (x) => x` or `([MyAttribute] x) => x`. I think we should require parens around the argument list if you're using attributes.

Reply:
@cston is there an open question in this doc for that?